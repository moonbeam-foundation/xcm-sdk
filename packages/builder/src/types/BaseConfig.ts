export interface BaseConfigConstructorParams {
  module: string;
  func: string;
}

export class BaseConfig {
  readonly module: string;

  readonly func: string;

  constructor({ module, func }: BaseConfigConstructorParams) {
    this.module = module;
    this.func = func;
  }
}
