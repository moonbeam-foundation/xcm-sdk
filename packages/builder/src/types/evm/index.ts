export * from './ContractConfig';
export * from './EvmQueryConfig';
