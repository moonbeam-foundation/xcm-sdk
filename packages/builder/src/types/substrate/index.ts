export * from './ExtrinsicConfig';
export * from './SubstrateCallConfig';
export * from './SubstrateQueryConfig';
