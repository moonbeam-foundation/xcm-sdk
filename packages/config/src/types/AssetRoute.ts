import type {
  AssetMinConfigBuilder,
  BalanceConfigBuilder,
  ContractConfigBuilder,
  ExtrinsicConfigBuilder,
  FeeConfigBuilder,
} from '@moonbeam-network/xcm-builder';
import type {
  AnyChain,
  Asset,
  ChainAsset,
  SetOptional,
} from '@moonbeam-network/xcm-types';

export interface AssetRouteConstructorParams {
  source: SourceConfig;
  destination: DestinationConfig;
  contract?: ContractConfigBuilder;
  extrinsic?: ExtrinsicConfigBuilder;
}

export interface SourceConfig {
  asset: Asset;
  chain: AnyChain;
  balance: BalanceConfigBuilder;
  fee: FeeConfig;
  destinationFee?: {
    asset?: Asset;
    balance: BalanceConfigBuilder;
  };
  min?: AssetMinConfigBuilder;
}

export interface DestinationConfig extends Omit<SourceConfig, 'fee'> {
  fee: DestinationFeeConfig;
}

export interface FeeConfig {
  asset: Asset;
  balance: BalanceConfigBuilder;
  // NOTE: Sometimes we need to add some extra amount ("XCM Delivery Fee") to a fee
  // that is returned by "paymentInfo" for extrinsic to not fail.
  extra?: number;
}

export interface DestinationFeeConfig
  extends SetOptional<FeeConfig, 'balance'> {
  amount: number | FeeConfigBuilder;
}

export class AssetRoute {
  readonly source: SourceConfig;

  readonly destination: DestinationConfig;

  readonly contract?: ContractConfigBuilder;

  readonly extrinsic?: ExtrinsicConfigBuilder;

  constructor({
    source,
    destination,
    contract,
    extrinsic,
  }: AssetRouteConstructorParams) {
    this.source = source;
    this.destination = destination;
    this.contract = contract;
    this.extrinsic = extrinsic;
  }

  getDestinationFeeAssetOnSource(): ChainAsset {
    return this.source.chain.getChainAsset(
      this.source.destinationFee?.asset || this.destination.fee.asset,
    );
  }
}
