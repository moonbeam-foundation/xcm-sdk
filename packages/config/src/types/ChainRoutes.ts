import type { AnyAsset, AnyChain, Asset } from '@moonbeam-network/xcm-types';
import { getKey } from '../config.utils';
import {
  AssetRoute,
  type AssetRouteConstructorParams,
  type SourceConfig,
} from './AssetRoute';

export interface ChainRoutesConstructorParams {
  chain: AnyChain;
  routes: RoutesParam[];
}

interface RoutesParam extends Omit<AssetRouteConstructorParams, 'source'> {
  source: Omit<SourceConfig, 'chain'>;
}

export class ChainRoutes {
  readonly chain: AnyChain;

  protected routes: Map<string, AssetRoute>;

  constructor({ chain, routes }: ChainRoutesConstructorParams) {
    this.chain = chain;
    this.routes = new Map(
      routes.map(({ source, destination, contract, extrinsic, monitoring }) => [
        `${source.asset.key}-${destination.chain.key}`,
        new AssetRoute({
          source: { ...source, chain },
          destination,
          contract,
          extrinsic,
          monitoring,
        }),
      ]),
    );
  }

  getRoutes(): AssetRoute[] {
    return Array.from(this.routes.values());
  }

  getAssetRoutes(keyOrAsset: string | AnyAsset): AssetRoute[] {
    const key = getKey(keyOrAsset);

    return this.getRoutes().filter((route) => route.source.asset.key === key);
  }

  getAssetDestinations(keyOrAsset: string | AnyAsset): AnyChain[] {
    return this.getAssetRoutes(keyOrAsset).map(
      (assetConfig) => assetConfig.destination.chain,
    );
  }

  getDestinationAssets(keyOrChain: string | AnyChain): Asset[] {
    const key = getKey(keyOrChain);

    return this.getRoutes()
      .filter((route) => route.destination.chain.key === key)
      .map((route) => route.source.asset);
  }

  getAssetRoute(
    asset: string | AnyAsset,
    destination: string | AnyChain,
  ): AssetRoute {
    const assetKey = getKey(asset);
    const destKey = getKey(destination);
    const route = this.routes.get(`${assetKey}-${destKey}`);

    if (!route) {
      throw new Error(
        `AssetRoute for asset ${assetKey} and destination ${destKey} not found`,
      );
    }

    return route;
  }
}
