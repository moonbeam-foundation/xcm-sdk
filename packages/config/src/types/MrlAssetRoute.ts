import type { BalanceConfigBuilder } from '@moonbeam-network/xcm-builder';
import type { Asset } from '@moonbeam-network/xcm-types';
import {
  AssetRoute,
  type AssetRouteConstructorParams,
  type MrlConfig,
  type SourceConfig,
} from './AssetRoute';

export interface MrlAssetRouteConstructorParams
  extends AssetRouteConstructorParams {
  source: MrlSourceConfig;
  mrl: MrlConfig;
}

export interface MrlSourceConfig extends SourceConfig {
  moonChainFee?: {
    asset: Asset;
    balance: BalanceConfigBuilder;
  };
}

export class MrlAssetRoute extends AssetRoute {
  readonly mrl: MrlConfig;
  readonly source: MrlSourceConfig;

  constructor({
    source,
    destination,
    contract,
    extrinsic,
    mrl,
  }: MrlAssetRouteConstructorParams & { source: MrlSourceConfig }) {
    super({ source, destination, contract, extrinsic });
    this.mrl = mrl;
    this.source = source;
  }
}
