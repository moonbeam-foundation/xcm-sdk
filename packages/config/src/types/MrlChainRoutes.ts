import type { AnyAsset, AnyChain } from '@moonbeam-network/xcm-types';
import { getKey } from '../config.utils';
import { ChainRoutes, type ChainRoutesConstructorParams } from './ChainRoutes';
import {
  MrlAssetRoute,
  type MrlAssetRouteConstructorParams,
  type MrlSourceConfig,
} from './MrlAssetRoute';

export interface MrlChainRoutesConstructorParams
  extends ChainRoutesConstructorParams {
  routes: MrlRoutesParam[];
}

interface MrlRoutesParam
  extends Omit<MrlAssetRouteConstructorParams, 'source'> {
  source: Omit<MrlSourceConfig, 'chain'>;
}

export class MrlChainRoutes extends ChainRoutes {
  readonly #routes: Map<string, MrlAssetRoute>;

  constructor({ chain, routes }: MrlChainRoutesConstructorParams) {
    super({ chain, routes });
    this.#routes = new Map(
      routes.map(({ source, destination, contract, extrinsic, mrl }) => [
        `${source.asset.key}-${destination.chain.key}`,
        new MrlAssetRoute({
          source: { ...source, chain },
          destination,
          contract,
          extrinsic,
          mrl,
        }),
      ]),
    );
  }

  getRoutes(): MrlAssetRoute[] {
    return Array.from(this.#routes.values());
  }

  getAssetRoute(
    asset: string | AnyAsset,
    destination: string | AnyChain,
  ): MrlAssetRoute {
    const assetKey = getKey(asset);
    const destKey = getKey(destination);
    const route = this.#routes.get(`${assetKey}-${destKey}`);

    if (!route) {
      throw new Error(
        `AssetRoute for asset ${assetKey} and destination ${destKey} not found`,
      );
    }

    return route;
  }
}
