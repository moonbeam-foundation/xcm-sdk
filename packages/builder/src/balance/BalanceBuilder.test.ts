import { describe, expect, it } from 'vitest';

import { TypeRegistry, U128 } from '@polkadot/types';
import type {
  FrameSystemAccountInfo,
  PalletBalancesAccountData,
} from '@polkadot/types/lookup';
import { testChainAsset } from '../../fixtures';
import type { SubstrateQueryConfig } from '../types/substrate/SubstrateQueryConfig';
import {
  BalanceBuilder,
  calculateSystemAccountBalance,
} from './BalanceBuilder';
import type { PalletBalancesAccountDataOld } from './BalanceBuilder.interfaces';

function balanceOf(number: number | string): U128 {
  return new U128(new TypeRegistry(), number);
}

/**
 * Using snapshot to test bigint values because jest does not support bigint
 */

describe('balanceBuilder', () => {
  const address = '<ADDRESS>';
  const asset = testChainAsset;

  describe('assets', () => {
    describe('account', () => {
      const config = BalanceBuilder()
        .substrate()
        .assets()
        .account()
        .build({ address, asset }) as SubstrateQueryConfig;

      it('should be correct config', () => {
        expect(config).toMatchSnapshot();
      });

      it('should transform correctly', async () => {
        await expect(
          config.transform({
            unwrapOrDefault: () => ({ balance: balanceOf(999) }),
          }),
        ).resolves.toMatchSnapshot();
      });
    });
  });

  describe('foreignAssets', () => {
    describe('account', () => {
      const config = BalanceBuilder()
        .substrate()
        .foreignAssets()
        .account()
        .build({
          address,
          asset,
        }) as SubstrateQueryConfig;

      it('should be correct config', () => {
        expect(config).toMatchSnapshot();
      });

      it('should transform correctly', async () => {
        await expect(
          config.transform({
            unwrapOrDefault: () => ({ balance: balanceOf(999) }),
          }),
        ).resolves.toMatchSnapshot();
      });
    });
  });

  describe('system', () => {
    describe('account', () => {
      const config = BalanceBuilder()
        .substrate()
        .system()
        .account()
        .build({ address, asset }) as SubstrateQueryConfig;

      it('should be correct config', () => {
        expect(config).toMatchSnapshot();
      });

      it('should transform correctly with frozen balance', async () => {
        await expect(
          config.transform({
            data: {
              free: balanceOf(999),
              frozen: balanceOf(99),
              reserved: balanceOf(0),
            },
          }),
        ).resolves.toMatchSnapshot();
      });

      it('should transform correctly with reserved balance', async () => {
        await expect(
          config.transform({
            data: {
              free: balanceOf(999),
              frozen: balanceOf(99),
              reserved: balanceOf(99),
            },
          }),
        ).resolves.toMatchSnapshot();
      });

      it('should transform correctly with miscFrozen balance', async () => {
        await expect(
          config.transform({
            data: {
              free: balanceOf(999),
              miscFrozen: balanceOf(99),
              reserved: balanceOf(0),
            },
          }),
        ).resolves.toMatchSnapshot();
      });
    });

    describe('accountEquilibrium', () => {
      const config = BalanceBuilder()
        .substrate()
        .system()
        .accountEquilibrium()
        .build({
          address,
          asset: testChainAsset.copyWith({
            ids: {
              balanceId: 25969,
            },
          }),
        }) as SubstrateQueryConfig;

      it('should be correct config', () => {
        expect(config).toMatchSnapshot();
      });

      it('should return 0 if response empty', async () => {
        await expect(
          config.transform({
            data: {
              isEmpty: true,
            },
          }),
        ).resolves.toMatchSnapshot();
      });

      it('should transform v0 correctly', async () => {
        await expect(
          config.transform({
            data: {
              isEmpty: false,
              toJSON: () => ({
                v0: {
                  balance: [
                    [
                      25969,
                      {
                        positive: 2954852364913,
                      },
                    ],
                    [
                      6648164,
                      {
                        positive: 18800000000,
                      },
                    ],
                    [
                      1735159154,
                      {
                        positive: 245805795,
                      },
                    ],
                  ],
                  lock: 0,
                },
              }),
            },
          }),
        ).resolves.toMatchSnapshot();
      });

      it('should transform correctly if not v0', async () => {
        await expect(
          config.transform({
            data: {
              isEmpty: false,
              toJSON: () => [
                [
                  25969,
                  {
                    positive: 62730076931,
                  },
                ],
                [
                  1735159154,
                  {
                    positive: 44972367,
                  },
                ],
              ],
            },
          }),
        ).resolves.toMatchSnapshot();
      });
    });
  });

  describe('tokens', () => {
    describe('accounts', () => {
      const config = BalanceBuilder()
        .substrate()
        .tokens()
        .accounts()
        .build({ address, asset }) as SubstrateQueryConfig;

      it('should be correct config', () => {
        expect(config).toMatchSnapshot();
      });

      it('should transform correctly', async () => {
        await expect(
          config.transform({
            free: balanceOf(999),
            frozen: balanceOf(99),
            reserved: balanceOf(0),
          }),
        ).resolves.toMatchSnapshot();
      });
    });
  });

  describe('calculateSystemAccountBalance', () => {
    it('should correctly calculate balance with PalletBalancesAccountData', async () => {
      const response = {
        data: {
          flags: balanceOf(0),
          free: balanceOf(1000),
          frozen: balanceOf(300),
          reserved: balanceOf(200),
        } as PalletBalancesAccountData,
      } as FrameSystemAccountInfo;

      const result = await calculateSystemAccountBalance(response);
      expect(result).toBe(BigInt(900));
    });

    it('should correctly calculate balance with PalletBalancesAccountDataOld', async () => {
      const response = {
        data: {
          feeFrozen: balanceOf(0),
          free: balanceOf(1000),
          miscFrozen: balanceOf(300),
          reserved: balanceOf(200),
        } as PalletBalancesAccountDataOld,
      } as unknown as FrameSystemAccountInfo;

      const result = await calculateSystemAccountBalance(response);
      expect(result).toBe(BigInt(900));
    });

    it('should handle when reserved is greater than frozen', async () => {
      const response = {
        data: {
          flags: balanceOf(0),
          free: balanceOf(1000),
          frozen: balanceOf(300),
          reserved: balanceOf(400),
        } as PalletBalancesAccountData,
      } as FrameSystemAccountInfo;

      const result = await calculateSystemAccountBalance(response);
      expect(result).toBe(BigInt(1000));
    });

    it('should return 0 when all balances are zero', async () => {
      const response = {
        data: {
          flags: balanceOf(0),
          free: balanceOf(0),
          frozen: balanceOf(0),
          reserved: balanceOf(0),
        } as PalletBalancesAccountData,
      } as FrameSystemAccountInfo;

      const result = await calculateSystemAccountBalance(response);
      expect(result).toBe(BigInt(0));
    });

    it('should handle large numbers correctly', async () => {
      const largeNumber = '123456789012345678901234567890';
      const response = {
        data: {
          flags: balanceOf(0),
          free: balanceOf(largeNumber),
          frozen: balanceOf(largeNumber),
          reserved: balanceOf(100),
        } as PalletBalancesAccountData,
      } as FrameSystemAccountInfo;

      const result = await calculateSystemAccountBalance(response);
      expect(result).toBe(
        BigInt(largeNumber) - (BigInt(largeNumber) - BigInt(100)),
      );
    });
  });
});
