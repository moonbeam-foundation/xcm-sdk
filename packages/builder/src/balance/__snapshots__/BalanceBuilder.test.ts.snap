// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`balanceBuilder > assets > account > should be correct config 1`] = `
SubstrateQueryConfig {
  "args": [
    "USDT",
    "<ADDRESS>",
  ],
  "func": "account",
  "module": "assets",
  "transform": [Function],
}
`;

exports[`balanceBuilder > assets > account > should transform correctly 1`] = `999n`;

exports[`balanceBuilder > foreignAssets > account > should be correct config 1`] = `
SubstrateQueryConfig {
  "args": [
    {
      "interior": {
        "X2": [
          {
            "GlobalConsensus": {
              "ethereum": {
                "chainId": 1,
              },
            },
          },
          {
            "AccountKey20": {
              "key": "0x98891e5FD24Ef33A488A47101F65D212Ff6E650E",
            },
          },
        ],
      },
      "parents": 2,
    },
    "<ADDRESS>",
  ],
  "func": "account",
  "module": "foreignAssets",
  "transform": [Function],
}
`;

exports[`balanceBuilder > foreignAssets > account > should transform correctly 1`] = `999n`;

exports[`balanceBuilder > system > account > should be correct config 1`] = `
SubstrateQueryConfig {
  "args": [
    "<ADDRESS>",
  ],
  "func": "account",
  "module": "system",
  "transform": [Function],
}
`;

exports[`balanceBuilder > system > account > should transform correctly with frozen balance 1`] = `900n`;

exports[`balanceBuilder > system > account > should transform correctly with miscFrozen balance 1`] = `900n`;

exports[`balanceBuilder > system > account > should transform correctly with reserved balance 1`] = `999n`;

exports[`balanceBuilder > system > accountEquilibrium > should be correct config 1`] = `
SubstrateQueryConfig {
  "args": [
    "<ADDRESS>",
  ],
  "func": "account",
  "module": "system",
  "transform": [Function],
}
`;

exports[`balanceBuilder > system > accountEquilibrium > should return 0 if response empty 1`] = `0n`;

exports[`balanceBuilder > system > accountEquilibrium > should transform correctly if not v0 1`] = `62730076931n`;

exports[`balanceBuilder > system > accountEquilibrium > should transform v0 correctly 1`] = `2954852364913n`;

exports[`balanceBuilder > tokens > accounts > should be correct config 1`] = `
SubstrateQueryConfig {
  "args": [
    "<ADDRESS>",
    "USDT",
  ],
  "func": "accounts",
  "module": "tokens",
  "transform": [Function],
}
`;

exports[`balanceBuilder > tokens > accounts > should transform correctly 1`] = `900n`;
